src/lib/index.ts

```diff
@@ -5,309 +5,311 @@ const UPDATE_ACTION = '@ngrx/store/update-reducers';
-    nextState = mergeReducer(nextState, rehydratedState, action);
-
-    nextState = reducer(nextState, action);
-
-    if (action.type !== INIT_ACTION) {
-      syncStateUpdate(
-        nextState,
-        stateKeys,
-        config.storage,
-        config.storageKeySerializer,
-        config.removeOnUndefined,
-        config.syncCondition,
-      );
+    if (config.storageKeySerializer === undefined) {
+        config.storageKeySerializer = (key) => key;
     }
 
-    return nextState;
-  };
+    if (config.restoreDates === undefined) {
+        config.restoreDates = true;
+    }
+
+    // Use default merge reducer.
+    let mergeReducer = config.mergeReducer;
+
+    if (mergeReducer === undefined || typeof mergeReducer !== 'function') {
+        mergeReducer = defaultMergeReducer;
+    }
+
+    const stateKeys = validateStateKeys(config.keys);
+    const rehydratedState = config.rehydrate
+        ? rehydrateApplicationState(stateKeys, config.storage, config.storageKeySerializer, config.restoreDates)
+        : undefined;
+
+    return function (state: any, action: any) {
+        let nextState: any;
+
+        // If state arrives undefined, we need to let it through the supplied reducer
+        // in order to get a complete state as defined by user
+        if (action.type === INIT_ACTION && !state) {
+            nextState = reducer(state, action);
+        } else {
+            nextState = { ...state };
+        }
+
+        // Merge the store state with the rehydrated state using
+        // either a user-defined reducer or the default.
+        nextState = mergeReducer(nextState, rehydratedState, action);
+
+        nextState = reducer(nextState, action);
+
+        if (action.type !== INIT_ACTION) {
+            syncStateUpdate(
+                nextState,
+                stateKeys,
+                config.storage,
+                config.storageKeySerializer,
+                config.removeOnUndefined,
+                config.syncCondition
+            );
+        }
+
+        return nextState;
+    };
 };
 
 export interface LocalStorageConfig {
-  keys: any[];
-  rehydrate?: boolean;
-  storage?: Storage;
-  removeOnUndefined?: boolean;
-  restoreDates?: boolean;
-  storageKeySerializer?: (key: string) => string;
-  syncCondition?: (state: any) => any;
-  checkStorageAvailability?: boolean;
-  mergeReducer?: (state: any, rehydratedState: any, action: any) => any;
+    keys: Keys;
+    rehydrate?: boolean;
+    storage?: Storage;
+    removeOnUndefined?: boolean;
+    restoreDates?: boolean;
+    storageKeySerializer?: (key: string) => string;
+    syncCondition?: (state: any) => any;
+    checkStorageAvailability?: boolean;
+    mergeReducer?: (state: any, rehydratedState: any, action: any) => any;
+}
+
+interface KeyConfiguration {
+    [key: string]: string[] | number[] | KeyConfiguration[] | Options | ((key: string, value: any) => any);
 }
+
+interface Options {
+    serialize?: (state: any) => any;
```

Comment:
This is missing the `deserialize` function.

Reply:
Added.